src/libraries/Microsoft.Extensions.Logging/src/LoggerFactoryScopeProvider.cs

```diff
@@ -48,12 +47,48 @@ void Report(Scope current)
                     }
 
                     callback(activityLogScope, state);
+
+                    // Tags and baggage are opt-in and thus we assume that most of the time it will not be used.
+                    if ((_activityTrackingOption & ActivityTrackingOptions.Tags) != 0)
+                    {
+                        const string additionalItemsTagsPropertyKey = "__ActivityTagItemsLogScope__";
```

Comment:
>additionalItemsTagsPropertyKey [](start = 37, length = 30)

You are caching the object inside the activity which is wrong because the Tags can be changed between logging calls. This is the whole point we couldn't cache it in ActivityLogScope.

Reply:
I only change the Scope-Wrapper. At enumeration time the tags/baggage will always be enumerated.

I wrote a test for this. See `TestActivityTrackingOptions_ShouldAddNewTagAndBaggageItemsAtRuntime_WhenTagsAndBaggageOptionAreSet`

The reason we can't put them into the other scope is that the scope is accessible with count and index. If the other scope would be also just IEnumerable it would have been possible to use it.

Did I miss something?

Reply:
> I only chance the Scope-Wrapper. At enumeration time the tags/baggage will always be enumerated.

I am not sure if this is true.

Imagin the following scenario: 

- Create and start Activity.
- add the Baggage <"key1", "value1">.
- start the logging and create the first logging scope `logger.BeginScope("Scope1")` then start logging some info.
- At that time, you'll call LogScopeForAdditionalActivityItems passing the Activity.Baggage (which will have <"key1", "value1"> entry). LogScopeForAdditionalActivityItems will create AdditionalActivityItemsLogScope object and cache it inside the Activity object. so far so good.
- after that try to add a new Baggage to the activity <"key2", "value2">.
- create a second scope `logger.BeginScope("Scope2")` then log some more info.
- at that time, you'll call LogScopeForAdditionalActivityItems again which will retrieve the stored AdditionalActivityItemsLogScope from the activity and try to log the baggage from this object which still has <"key1", "value1"> only and that means <"key2", "value2"> will be missing at that time in the logging.

Let me know if I am missing anything here.

Reply:
Do you talk about a child `Activity` rather than a log scope? AFAIK a log scope has nothing to do with scoping the values. The log scope just advises the logger, which is called within this Scope, to log the values (represented by this `Scope`).

Let's see if I got your statements right. Does the following code reflect what you think?

```
Activity activity = new Activity("ScopeActivity");
activity.Start();

activity.AddTag("MyTagKey1", "1");
activity.AddBaggage("MyBaggageKey1", "1");
logger.LogInformation("Message1");

using (logger.BeginScope("Scope1"))
{
    activity.AddTag("MyTagKey2", "2");
    activity.AddBaggage("MyBaggageKey2", "2");
    logger.LogInformation("Message2");

    using (logger.BeginScope("Scope2"))
    {
        activity.AddTag("MyTagKey3", "3");
        activity.AddBaggage("MyBaggageKey3", "3");

        using (logger.BeginScope("Scope3"))
        {
            activity.SetTag("MyTagKey3", "4");
            activity.SetBaggage("MyBaggageKey3", "4");
        }
    }

    logger.LogInformation("Message4");

    activity.Stop();
}
```

![image](https://user-images.githubusercontent.com/7792779/110379870-a8f3eb80-8057-11eb-8ce6-bb4915241c94.png)

Reply:
I added this as test: `TestActivityTrackingOptions_ShouldAddNewTagAndBaggageItemsAtRuntime_WhenTagsAndBaggageOptionAreSetAndWithNestedScopes`